src/http/error_handler.ml

```diff
@@ -189,18 +189,22 @@ let customize template (error : Error.error) =
 
 
 
-(* TODO LATER Make a nice default template. *)
 let default_template debug_dump response =
-  let response =
+  let status = Dream.status response in
+  let code = Dream.status_to_int status
+  and reason = Dream.status_to_string status in
+  let body =
     match debug_dump with
-    | None -> response
-    | Some info ->
-      response
-      |> Dream.with_body info
-      |> Dream.with_header "Content-Type" "text/plain"
+    | Some s ->
+      let debug_dump = Dream__pure.Formats.html_escape s in
+      Fallback_template.render_debug ~debug_dump ~code ~reason
+    | None ->
+      Fallback_template.render ~code ~reason
```

Comment:
This changes the behavior described in [*Errors*](https://aantron.github.io/dream/#errors):

> The default error handler logs errors and its template generates completely empty responses, to avoid internationalization issues.

Without `~debug:true`, Dream should not assume any natural language or any style of writing or visual presentation, and leave that completely up to Dream's user (or a higher-level framework or project generator). Otherwise, it adds a new thing that people have to remember to "tuck in" to avoid leaking non-internationalized strings, dealing with style issues, etc.

Conveniently, at least Chromium-based browsers already internationalize their built-in error pages for empty responses, so "base" Dream is piggy-backing on that.

Reply:
Agree! I'm wondering why the error handling is not provided as a middleware then? It seems to offer the same intended flexibility to the user, while re-using existing concepts instead of another semantics?

I'm guessing this is because of how Dream handles errors internally, so maybe that's simply not possible?

Reply:
> I'm wondering why the error handling is not provided as a middleware then? It seems to offer the same intended flexibility to the user, while re-using existing concepts instead of another semantics?

Error handlers are indeed very middleware-like. But middleware-based error handlers are an (1) opt-in way to (2) handle errors from *within* your app. Dream's error handler handles errors from the app and from the low-level libraries, to capture errors from all levels and pass them through a single handler as a filter. That also means that a default error handler must always be present, because low-level errors (such as 400 Bad Request responses) get generated by the libraries under Dream no matter what an app does or does not do. It would be strange if a 400 Bad Request coming from deep inside http/af got somehow routed into a middleware halfway through a user app's middleware stack for postprocessing by a template. The other option, to have the framework simply emit the low-level 400 Bad Request, leaks undecorated, untranslated strings to the client, that are under the control of http/af, etc., and not under the control of the framework user, the Web app developer.

If there is a TLS handshake error, how would an advanced user gain control over its logging, if error handlers are middlewares?

There is also the issue of what happens when the error handler itself raises an error. If error handlers are hidden in middleware stacks, do we know if any given error is a real error, or a "double fault" coming from a failed error handler? If the framework distinguishes the error handler, it can detect double faults exactly in an easy way, at the call site.

So error handlers are some kind of middleware-like built-in thing that have some slightly different semantics.



> I'm guessing this is because of how Dream handles errors internally, so maybe that's simply not possible?

No, it's not a limitation. Dream could offer middleware-based error handlers as the default, but they would have all the drawbacks above. A user can still implement a middleware-based error handler trivially for their own purposes, if they need to deal with app-only errors in a way that middleware-based error handlers are well-suited for.

Reply:
(By the way, there is a built-in error handling middleware, [`Dream.catch`](https://aantron.github.io/dream/#val-catch), but its job is just to forward app errors to the "big central error handler" that is also catching http/af errors, TLS errors, etc. This middleware can be [removed](https://aantron.github.io/dream/#builtin) to suppress its behavior, for completely custom app error handling. Lower-level errors still won't go into the app, but to the big error handler.)